### PR TITLE
Swap and vesting bonus

### PR DESCRIPTION
**Pull request by lovesh**
Signed-off-by: lovesh <lovesh.bond@gmail.com>

### CODE REVIEW

File: pallets/token_migration/src/lib.rs [hunk 1]
```diff
@@ -198,17 +314,299 @@ impl<T: Trait> Module<T> {
...
+                        T::BlockNumberToBalance::convert(milestones_passed) * bonus_per_milestone;
+                    let expected_locked = total_bonus.saturating_sub(bonus_to_be_unlocked_till_now);
+                    if expected_locked < locked_bonus {
+                        bonus_to_unlock += locked_bonus - expected_locked;
+                        bonuses[i].1 = expected_locked
+                    }
+                }
+            }
+            i += 1;
+        }
```
Comment (bddap) [inline pallets/token_migration/src/lib.rs:517]: This block gets a bit hairy. Want me to attempt to refactor for readability?

**Review by bddap — APPROVED** [review]
This module is too complex for me to tell with confidence that its correct. If we require that level of verification I'm happy to work with you to simplify it.

The module is well organized. I don't see any bugs and haven't detected any code smell.

File: pallets/token_migration/src/lib.rs [hunk 2]
```diff
@@ -198,17 +341,320 @@ impl<T: Trait> Module<T> {
...
+            acc,
+            bonus.total_locked_bonus,
+            WithdrawReasons::all(),
+        );
+        Bonuses::<T>::insert(acc.clone(), bonus);
+    }
+
+    /// Update storage and lock for bonus of an account after a claim has been made. Remove lock and
+    /// storage entry if all bonus claimed else reset lock and update storage.
+    fn update_bonus_after_claim(acc: &T::AccountId, bonus: Bonus<BalanceOf<T>, T::BlockNumber>) {
```
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/token_migration/src/lib.rs:558]: Yes, will combine.

File: pallets/token_migration/src/lib.rs (hunk 1 above)
Comment (lovesh) [reply pallets/token_migration/src/lib.rs:517]: Does hairy mean long? If not, can you elaborate?

Comment (bddap) [reply pallets/token_migration/src/lib.rs:517]: The nested branches and imperative style makes intent difficult to discern. And yes, it is a bit long.

Comment (lovesh) [reply pallets/token_migration/src/lib.rs:517]: Removed an extra branch. Added note about why it was there.

**Pull request merged**